src/serde_helpers.rs

```diff
@@ -59,6 +64,87 @@ pub fn serialize_u64_as_i64<S: Serializer>(val: &u64, serializer: S) -> Result<S
     }
 }
 
+/// Contains functions to serialize a u32 as an f64 (BSON double) and deserialize a
+/// u32 from an f64 (BSON double).
+///
+/// ```rust
+/// # use serde::{Serialize, Deserialize};
+/// # use bson::serde_helpers::u32_as_f64;
+/// #[derive(Serialize, Deserialize)]
+/// struct FileInfo {
+///     #[serde(with = "u32_as_f64")]
+///     pub size_bytes: u32,
+/// }
+/// ```
+pub mod u32_as_f64 {
+    use serde::{de, Deserialize, Deserializer, Serializer};
+
+    /// Deserializes a u32 from an f64 (BSON double). Errors if an exact conversion is not possible.
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<u32, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let f = f64::deserialize(deserializer)?;
+        if (f - f as u32 as f64).abs() <= f64::EPSILON {
+            Ok(f as u32)
+        } else {
+            Err(de::Error::custom(format!(
+                "cannot convert f64 (BSON double) {} to u32",
+                f
+            )))
+        }
+    }
+
+    /// Serializes a u32 as an f64 (BSON double).
+    pub fn serialize<S: Serializer>(val: &u32, serializer: S) -> Result<S::Ok, S::Error> {
+        serializer.serialize_f64(*val as f64)
+    }
+}
+
+/// Contains functions to serialize a u64 as an f64 (BSON double) and deserialize a
+/// u64 from an f64 (BSON double).
+///
+/// ```rust
+/// # use serde::{Serialize, Deserialize};
+/// # use bson::serde_helpers::u64_as_f64;
+/// #[derive(Serialize, Deserialize)]
+/// struct FileInfo {
+///     #[serde(with = "u64_as_f64")]
+///     pub size_bytes: u64,
+/// }
+/// ```
+pub mod u64_as_f64 {
+    use serde::{de, ser, Deserialize, Deserializer, Serializer};
+
+    /// Deserializes a u64 from an f64 (BSON double). Errors if an exact conversion is not possible.
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<u64, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let f = f64::deserialize(deserializer)?;
+        if (f - f as u64 as f64).abs() <= f64::EPSILON {
+            Ok(f as u64)
+        } else {
+            Err(de::Error::custom(format!(
+                "cannot convert f64 (BSON double) {} to u64",
+                f
+            )))
+        }
+    }
+
+    /// Serializes a u64 as an f64 (BSON double). Errors if an exact conversion is not possible.
+    pub fn serialize<S: Serializer>(val: &u64, serializer: S) -> Result<S::Ok, S::Error> {
+        if val < &u64::MAX && *val == *val as f64 as u64 {
```

Comment:
The comparison to `u64::MAX` is used here since `val as f64 as u64` works for `u64::MAX` even though the `f64` created by it is not equivalent to `u64::MAX`:
```
u64::MAX: 18446744073709551615
u64::MAX as f64: 18446744073709552000
u64::MAX as f64 as u64: 18446744073709551615
```